### ISSUE TITLE
StatsModels compatibility post-Terms 2.0: Son of Terms

### ISSUE THREAD

**Issue by kleinschmidt** [JuliaInterop/RCall.jl#306]
Terms 2.0 (Son of Terms) (JuliaStats/StatsModels.jl#71) changes the representation of the formula rather dramatically; much of the information that was carried in the `Terms` struct is now present in the `FormulaTerm` that's created by `@formula`.  This somewhat complicates copying those objects to and from R, since the R formula is just a quoted expression.  As long as we're okay with formulae not always being fully "round trippable" then it should be possible to define sensible behavior that converts a `FormulaTerm` into an AST and (and use `@eval @formula $rformula` to do the inverse).

cc @dmbates

**Comment by dmbates** [comment]
@kleinschmidt I'm running into a bit of a problem rewriting the method `rcopy(::Type{Formula}, l::Ptr{LangSxp})`  I can create the expression `ex_orig` but I am not sure how to do the equivalent of a call to the `@formula` macro on it.  In other words, I want to, within a function, take an `Expr` that has the form of a formula and create a `FormulaTerm` from it.

If I look at the code in `StatsModels` it seems that after the checks on the first argument being `:~` and the length of the arguments being 3 then I should call something like 
```
StatsModels.terms!(StatsModels.sort_terms!(StatsModels.parse!(ex_orig)))
```
but that doesn't return the type of object I want.
```
julia> StatsModels.terms!(StatsModels.sort_terms!(StatsModels.parse!(:(y ~ 1 + x))))
:(($(Expr(:escape, :~)))(Term(:y), ($(Expr(:escape, :+)))(ConstantTerm(1), Term(:x))))

julia> @formula(y ~ 1 + x)
FormulaTerm
Response:
  y(unknown)
Predictors:
  1
  x(unknown)
```

I am perhaps making this much more complicated than it needs to be.  Can you provide some guidance?

**Comment by kleinschmidt** [comment]
You'll need to `eval` the expression returned by that, and apparently if you try to do that directly it chokes (because there's a lot of `esc` sprinkled into the transformed expression so it only works in a macro I guess?).  Probably the easiest thing to do is to just use the `@formula` macro:

```
julia> ex = :(y ~ 1 + x)
:(y ~ 1 + x)

julia> @eval @formula($ex)
FormulaTerm
Response:
  y(unknown)
Predictors:
  1
  x(unknown)
```

**Comment by dmbates** [comment]
@kleinschmidt Your fix works, thanks.

Going the other way, I want to take something like a `FormulaTerm` and create an expression that looks like a call to the binary operator `:~` .  In previous versions of `StatsModels` the expression passed to `@formula` was saved using
```
        ex_orig = Meta.quot(copy(ex))
```
Providing the original expression as part of the `Formula` object also allowed for the formula to be displayed in a compact form when `show`ing a model.  I encounter models where crossing (the `*` operator) causes a compact expression to be expanded to many, many terms and have no good way of getting back to the original form when displaying the formula.

Do you think it would be possible/advisable to cache a copy of the original expression in the result of a call to `@formula`?

**Comment by kleinschmidt** [comment]
I think it should be _possible_ (e.g. by adding a `ex_orig` field on the `FormulaTerm` struct) but it offends my sense of elegance a little bit since it would require an additional step in the `@formula` macro to lower the top-level call to `~` to a call to the `FormulaTerm` constructor instead of using polymorphism on the `~` function to create that magically when the arguments are `AbstractTerms`.  Can you open an issue on StatsModels to discuss?

**Comment by proudindiv** [comment]
StatsModels v0.6.0 (just updated) breaks RCall.   (It also broke Bootstrap StatsKit in it's using Bootstrap.)

```
using RCall;
┌ Info: Precompiling RCall [6f49c342-dc21-5d91-9882-a32aef131414]
└ @ Base loading.jl:1186
WARNING: could not import StatsModels.Formula into RCall
ERROR: LoadError: LoadError: UndefVarError: Formula not defined
Stacktrace:
 [1] top-level scope at none:0
 [2] include at ./boot.jl:326 [inlined]
 [3] include_relative(::Module, ::String) at ./loading.jl:1038
 [4] include at ./sysimg.jl:29 [inlined]
 [5] include(::String) at /home/gary/.julia/packages/RCall/iojZI/src/RCall.jl:2
 [6] top-level scope at none:0
 [7] include at ./boot.jl:326 [inlined]
 [8] include_relative(::Module, ::String) at ./loading.jl:1038
 [9] include(::Module, ::String) at ./sysimg.jl:29
 [10] top-level scope at none:2
 [11] eval at ./boot.jl:328 [inlined]
 [12] eval(::Expr) at ./client.jl:404
 [13] top-level scope at ./none:3
in expression starting at /home/gary/.julia/packages/RCall/iojZI/src/convert/formula.jl:39
in expression starting at /home/gary/.julia/packages/RCall/iojZI/src/RCall.jl:46
```

**Comment by kleinschmidt** [comment]
As a short term fix, you can pin statsmodels at 0.5 with `] pin StatsModels@0.5`.  I'm working on getting the appropriate version caps in place.

**Comment by palday** [comment]
I think #317 fixed this. Anybody care to close, @randy3k @simonbyrne @dmbates ?
